File: dotnet/Pragmastat.UnitTests/InvarianceTests.cs
using System.Diagnostics.CodeAnalysis;
using Pragmastat.Core;
using Pragmastat.Core.Metrology;
using Pragmastat.Distributions;
using Pragmastat.Distributions.Helpers;
using static Pragmastat.Toolkit;

namespace Pragmastat.UnitTests;

[SuppressMessage("ReSharper", "ConvertClosureToMethodGroup")]
public class InvarianceTests
{
    private const int Seed = 1729;

    private readonly int[] sampleSizes = [2, 3, 4, 5, 6, 7, 8, 9, 10];

    private void PerformTest(Func<Sample, Measurement> expression1, Func<Sample, Measurement> expression2)
    {
        var generator = UniformDistribution.Standard.Random(Seed);
        foreach (var n in sampleSizes)
        {
            var x = new Sample(generator.Next(n));
            var result1 = expression1(x);
            var result2 = expression2(x);
            Assert.Equal(result1.Unit.FullName, result2.Unit.FullName);
            Assert.Equal(result1.NominalValue, result2.NominalValue, 9);
        }
    }

    private void PerformTest(
        Func<Sample, Sample, Measurement> expression1,
        Func<Sample, Sample, Measurement> expression2)
    {
        var generator = UniformDistribution.Standard.Random(Seed);
        foreach (var n in sampleSizes)
        {
            var x = new Sample(generator.Next(n));
            var y = new Sample(generator.Next(n));
            var result1 = expression1(x, y);
            var result2 = expression2(x, y);
            Assert.Equal(result1.Unit.FullName, result2.Unit.FullName);
            Assert.Equal(result1.NominalValue, result2.NominalValue, 9);
        }
    }

    // Center

    [Fact]
    public void CenterShift() => PerformTest(x => Center(x + 2), x => Center(x) + 2);

    [Fact]
    public void CenterScale() => PerformTest(x => Center(2 * x), x => 2 * Center(x));

    [Fact]
    public void CenterNegate() => PerformTest(x => Center(-1 * x), x => -1 * Center(x));

    // Spread

    [Fact]
    public void SpreadShift() => PerformTest(x => Spread(x + 2), x => Spread(x));

    [Fact]
    public void SpreadScale() => PerformTest(x => Spread(2 * x), x => 2 * Spread(x));

    [Fact]
    public void SpreadNegate() => PerformTest(x => Spread(-1 * x), x => Spread(x));

    // Volatility

    [Fact]
    public void VolatilityScale() => PerformTest(x => Volatility(2 * x), x => Volatility(x));

    // Precision

    [Fact]
    public void PrecisionShift() => PerformTest(x => Precision(x + 2), x => Precision(x));

    [Fact]
    public void PrecisionScale() => PerformTest(x => Precision(2 * x), x => 2 * Precision(x));

    [Fact]
    public void PrecisionScaleNegate() => PerformTest(x => Precision(-2 * x), x => 2 * Precision(x));

    // MedShift

    [Fact]
    public void MedShiftShift() => PerformTest((x, y) => MedShift(x + 3, y + 2), (x, y) => MedShift(x, y) + 1);

    [Fact]
    public void MedShiftScale() => PerformTest((x, y) => MedShift(2 * x, 2 * y), (x, y) => 2 * MedShift(x, y));

    [Fact]
    public void MedShiftAntisymmetry() => PerformTest((x, y) => MedShift(x, y), (x, y) => -1 * MedShift(y, x));

    // MedRatio

    [Fact]
    public void MedRatioScale() => PerformTest((x, y) => MedRatio(2 * x, 3 * y), (x, y) => 2.0 / 3 * MedRatio(x, y));

    // MedSpread

    [Fact]
    public void MedSpreadEqual() => PerformTest(x => MedSpread(x, x), x => Spread(x));

    [Fact]
    public void MedSpreadSymmetry() => PerformTest((x, y) => MedSpread(x, y), (x, y) => MedSpread(y, x));

    [Fact]
    public void MedSpreadAverage() => PerformTest(x => MedSpread(x, 5 * x), x => 3 * Spread(x));

    [Fact]
    public void MedSpreadScale() => PerformTest((x, y) => MedSpread(-2 * x, -2 * y), (x, y) => 2 * MedSpread(x, y));

    // MedDisparity

    [Fact]
    public void MedDisparityShift() => PerformTest((x, y) => MedDisparity(x + 2, y + 2), (x, y) => MedDisparity(x, y));

    [Fact]
    public void MedDisparityScale() => PerformTest((x, y) => MedDisparity(2 * x, 2 * y), (x, y) => MedDisparity(x, y));

    [Fact]
    public void MedDisparityScaleNeg() => PerformTest((x, y) => MedDisparity(-2 * x, -2 * y), (x, y) => -1 * MedDisparity(x, y));

    [Fact]
    public void MedDisparityAntisymmetry() => PerformTest((x, y) => MedDisparity(x, y), (x, y) => -1 * MedDisparity(y, x));
}

File: rust/pragmastat/tests/invariance_tests.rs
use float_cmp::approx_eq;
use pragmastat::*;

/// Tests mathematical invariance properties of the estimators
#[cfg(test)]
mod invariance_tests {
    use super::*;

    const SEED: u64 = 1729;
    const SAMPLE_SIZES: [usize; 9] = [2, 3, 4, 5, 6, 7, 8, 9, 10];
    const TOLERANCE: f64 = 1e-9;

    /// Simple linear congruential generator for reproducible random numbers
    struct SimpleRng {
        state: u64,
    }

    impl SimpleRng {
        fn new(seed: u64) -> Self {
            SimpleRng { state: seed }
        }

        fn next_f64(&mut self) -> f64 {
            // Simple LCG parameters
            self.state = (self.state.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7fffffff;
            self.state as f64 / 0x7fffffff as f64
        }

        fn next_vec(&mut self, n: usize) -> Vec<f64> {
            (0..n).map(|_| self.next_f64()).collect()
        }
    }

    fn perform_test_one<F1, F2>(expr1: F1, expr2: F2)
    where
        F1: Fn(&[f64]) -> f64,
        F2: Fn(&[f64]) -> f64,
    {
        let mut rng = SimpleRng::new(SEED);
        for &n in &SAMPLE_SIZES {
            let x = rng.next_vec(n);
            let result1 = expr1(&x);
            let result2 = expr2(&x);
            assert!(
                approx_eq!(f64, result1, result2, epsilon = TOLERANCE),
                "Failed for n={}: {} != {}",
                n,
                result1,
                result2
            );
        }
    }

    fn perform_test_two<F1, F2>(expr1: F1, expr2: F2)
    where
        F1: Fn(&[f64], &[f64]) -> f64,
        F2: Fn(&[f64], &[f64]) -> f64,
    {
        let mut rng = SimpleRng::new(SEED);
        for &n in &SAMPLE_SIZES {
            let x = rng.next_vec(n);
            let y = rng.next_vec(n);
            let result1 = expr1(&x, &y);
            let result2 = expr2(&x, &y);
            assert!(
                approx_eq!(f64, result1, result2, epsilon = TOLERANCE),
                "Failed for n={}: {} != {}",
                n,
                result1,
                result2
            );
        }
    }

    // Helper functions for vector operations
    fn vec_add_scalar(v: &[f64], scalar: f64) -> Vec<f64> {
        v.iter().map(|&x| x + scalar).collect()
    }

    fn vec_mul_scalar(v: &[f64], scalar: f64) -> Vec<f64> {
        v.iter().map(|&x| x * scalar).collect()
    }

    // Center invariance tests

    #[test]
    fn center_shift() {
        perform_test_one(
            |x| center(&vec_add_scalar(x, 2.0)).unwrap(),
            |x| center(x).unwrap() + 2.0,
        );
    }

    #[test]
    fn center_scale() {
        perform_test_one(
            |x| center(&vec_mul_scalar(x, 2.0)).unwrap(),
            |x| 2.0 * center(x).unwrap(),
        );
    }

    #[test]
    fn center_negate() {
        perform_test_one(
            |x| center(&vec_mul_scalar(x, -1.0)).unwrap(),
            |x| -1.0 * center(x).unwrap(),
        );
    }

    // Spread invariance tests

    #[test]
    fn spread_shift() {
        perform_test_one(
            |x| spread(&vec_add_scalar(x, 2.0)).unwrap(),
            |x| spread(x).unwrap(),
        );
    }

    #[test]
    fn spread_scale() {
        perform_test_one(
            |x| spread(&vec_mul_scalar(x, 2.0)).unwrap(),
            |x| 2.0 * spread(x).unwrap(),
        );
    }

    #[test]
    fn spread_negate() {
        perform_test_one(
            |x| spread(&vec_mul_scalar(x, -1.0)).unwrap(),
            |x| spread(x).unwrap(),
        );
    }

    // Volatility invariance tests

    #[test]
    fn volatility_scale() {
        perform_test_one(
            |x| volatility(&vec_mul_scalar(x, 2.0)).unwrap(),
            |x| volatility(x).unwrap(),
        );
    }

    // Precision invariance tests

    #[test]
    fn precision_shift() {
        perform_test_one(
            |x| precision(&vec_add_scalar(x, 2.0)).unwrap(),
            |x| precision(x).unwrap(),
        );
    }

    #[test]
    fn precision_scale() {
        perform_test_one(
            |x| precision(&vec_mul_scalar(x, 2.0)).unwrap(),
            |x| 2.0 * precision(x).unwrap(),
        );
    }

    #[test]
    fn precision_scale_negate() {
        perform_test_one(
            |x| precision(&vec_mul_scalar(x, -2.0)).unwrap(),
            |x| 2.0 * precision(x).unwrap(),
        );
    }

    // MedShift invariance tests

    #[test]
    fn med_shift_shift() {
        perform_test_two(
            |x, y| med_shift(&vec_add_scalar(x, 3.0), &vec_add_scalar(y, 2.0)).unwrap(),
            |x, y| med_shift(x, y).unwrap() + 1.0,
        );
    }

    #[test]
    fn med_shift_scale() {
        perform_test_two(
            |x, y| med_shift(&vec_mul_scalar(x, 2.0), &vec_mul_scalar(y, 2.0)).unwrap(),
            |x, y| 2.0 * med_shift(x, y).unwrap(),
        );
    }

    #[test]
    fn med_shift_antisymmetry() {
        perform_test_two(
            |x, y| med_shift(x, y).unwrap(),
            |x, y| -1.0 * med_shift(y, x).unwrap(),
        );
    }

    // MedRatio invariance tests

    #[test]
    fn med_ratio_scale() {
        perform_test_two(
            |x, y| med_ratio(&vec_mul_scalar(x, 2.0), &vec_mul_scalar(y, 3.0)).unwrap(),
            |x, y| (2.0 / 3.0) * med_ratio(x, y).unwrap(),
        );
    }

    // MedSpread invariance tests

    #[test]
    fn med_spread_equal() {
        perform_test_one(|x| med_spread(x, x).unwrap(), |x| spread(x).unwrap());
    }

    #[test]
    fn med_spread_symmetry() {
        perform_test_two(
            |x, y| med_spread(x, y).unwrap(),
            |x, y| med_spread(y, x).unwrap(),
        );
    }

    #[test]
    fn med_spread_average() {
        perform_test_one(
            |x| med_spread(x, &vec_mul_scalar(x, 5.0)).unwrap(),
            |x| 3.0 * spread(x).unwrap(),
        );
    }

    #[test]
    fn med_spread_scale() {
        perform_test_two(
            |x, y| med_spread(&vec_mul_scalar(x, -2.0), &vec_mul_scalar(y, -2.0)).unwrap(),
            |x, y| 2.0 * med_spread(x, y).unwrap(),
        );
    }

    // MedDisparity invariance tests

    #[test]
    fn med_disparity_shift() {
        perform_test_two(
            |x, y| med_disparity(&vec_add_scalar(x, 2.0), &vec_add_scalar(y, 2.0)).unwrap(),
            |x, y| med_disparity(x, y).unwrap(),
        );
    }

    #[test]
    fn med_disparity_scale() {
        perform_test_two(
            |x, y| med_disparity(&vec_mul_scalar(x, 2.0), &vec_mul_scalar(y, 2.0)).unwrap(),
            |x, y| med_disparity(x, y).unwrap(),
        );
    }

    #[test]
    fn med_disparity_scale_neg() {
        perform_test_two(
            |x, y| med_disparity(&vec_mul_scalar(x, -2.0), &vec_mul_scalar(y, -2.0)).unwrap(),
            |x, y| -1.0 * med_disparity(x, y).unwrap(),
        );
    }

    #[test]
    fn med_disparity_antisymmetry() {
        perform_test_two(
            |x, y| med_disparity(x, y).unwrap(),
            |x, y| -1.0 * med_disparity(y, x).unwrap(),
        );
    }
}
